LibTest/io/exitCode/exitCode_A03_t02.dart

```diff
@@ -19,11 +19,13 @@ run_main() async {
   String executable = Platform.resolvedExecutable;
   String eScript = Platform.script.toString();
   int called = 0;
+  asyncStart();
   await Process.run(
           executable, [...Platform.executableArguments, eScript, "run"])
       .then((ProcessResult results) {
     Expect.equals(0, results.exitCode);
     called++;
+    asyncEnd();
```

Comment:
No, it's Ok to place it there. The idea is to make sure that async method was called and only once. For example
```dart
import "dart:async";
import "../../Utils/expect.dart";

main() async {
  asyncStart();
  await Future<int>.delayed(Duration(seconds: 1), () {return 42;}).then((value) {
    print(value);
    asyncEnd(); // Ok. It works, no matter that anything is printed after unittest-suite-success
    print(0);
  });
}
```
The test above will fail if, for some reason, `Future.then()` won't be called (then test runner determines `asyncStart()` without `asyncEnd()`) or will be called more than once (number of `asyncEnd()` calls should be equal the number of `asyncStart()` calls).

Reply:
I'm not sure that's the correct usage of `asyncEnd`. The test controller that runs in the browser declares a pass when it sees a `unittest-suite-success`: https://github.com/dart-lang/sdk/blob/12e52d8fde254a0aad0102223324f7aa953146c3/pkg/test_runner/lib/src/test_controller.js#L242-L244

So if you call `asyncEnd` and then do other checks, those checks will potentially run *after* the test result is sent back to the test driver and the test result is accepted as "pass".

Reply:
We discussed this with @mkustermann. The key points are that (1) there should be no assertions after the last `asyncEnd` (when the async task counter hits), and (2) `asyncEnd` should never be called when there are more `asyncStart`s to be run in the program.

As an example to (1), here we can't call `asyncEnd` in the awaited future:

```dart
main() async {
    asyncStart();
    await Future<int>.delayed(Duration(seconds: 1), () {
      ... // We can't call `asyncEnd` here
    });
    Expect.equals(...);
    asyncEnd();
}
```

This is because the part after the `await` will run after the future completes, so if we call `asyncEnd` in the future that concludes the test before the rest of the `main` runs.

If we didn't have `await`:

```dart
main() async {
    asyncStart();
    Future<int>.delayed(Duration(seconds: 1), () { // no `await` here
      ...
      asyncEnd();
      return 123;
    });
    Expect.equals(...);
}
```

Now we need to call `asyncEnd` in the future because the future will complete after `main` returns (as `main` does not wait for the future to complete).

As an example to (2), this is a bug:

```dart
main() async {
  asyncStart();
  test1();
  asyncEnd(); // BUG: the test result will be reported here.

  asyncStart();
  test2();
  asyncEnd();
}
```

Here the bug is that the first `asyncEnd()` decrements the async task counter to 0 and prints 'unittest-suite-success', which again concludes the test early.

For this particular test (1) applies, the `asyncEnd` should be after the `Expect.equals` in `main`.

Reply:
@osa1 and @mkustermann, it's great that this topic gets clarified, thanks!

A few things that I noticed in language tests using `asyncStart(...)` and `asyncEnd()` and from looking at the declarations of `asyncStart` and `asyncEnd`:

`asyncStart(k)` can be used to get the same effect as calling `asyncStart()` k times; this is never used.

Dynamically, the invocations of `asyncStart()` and `asyncEnd()` must determine a correct parenthesis structure. That is, if we use `(` to indicate that `asyncStart()` has been invoked and `)` to indicate `asyncEnd()`, we're allowed to have invocation histories like '()' or '((()))' or '(()()()(()))', but not ')' or '(()()))('. In other words, for every `asyncEnd()` there must be a corresponding `asyncStart()` in the history, and the total execution history must have a corresponding `asyncEnd()` for every `asyncStart()`. The final `asyncEnd()` is special in that it posts the 'unittest-suite-success' message which is interpreted by `processMessage` in 'test_controller.js'.

A common pattern in tests is to have an outer pair of `asyncStart/asyncEnd` invocations, and then possible a bunch of inner pairs. E.g., in 'language/await/await_test.dart':

```dart
main() {
  asyncStart();
  for (int i = 0; i < 11; i++) {
    asyncTest(staticMembers);
    ...
  }
  asyncEnd();
}

```

The `asyncStart()` is typically performed directly at the top level (here: of the body of `main`), and the `asyncEnd()` is usually performed in a callback, e.g., in 'language/await/postfix_expr_test.dart':

```dart
main() {
  asyncStart();
  test().then((_) { asyncEnd(); });
}
```

In the case where `asyncEnd()` is invoked at the top level it is guarded by a synchronization command (like `await completer.future` in 'async_star/async_star_cancel_test.dart'). In any case, some amount of synchronization may be required in order to ensure that the parenthesis structure is correct at run time, and also to ensure that the final `asyncEnd()` occurs after all elements of the test have been performed.

It may seem more visually appealing to have `asyncStart()` and `asyncEnd()` occur symmetrically: They both occur in the body of the same function, with the same amount of indentation. However, the approach where `asyncEnd()` is called in an argument to `.then` of a future may be simpler because there is no need to write extra code to achieve the required synchronization. So the style from 'postfix_expr_test.dart' could be a useful standard.

When we have a top-level `asyncStart/asyncEnd` pair (say, in `main`), we're free to have nested pairs, as shown in @mkustermann's example in the comment on 'element_type_A02_t03.dart'. This is like '(())'; however, as @mkustermann mentions, we're actually running the `asyncStart()` in `main`, then the `asyncStart()` in `isRuntime...`, then the `asyncEnd()` in `main` and finally `asyncEnd()` in `isRuntime...`. In other words, the parenthesis structure of the dynamic invocation sequence exists, but it doesn't reflect the syntactic location of the invocations of `asyncStart` and `asyncEnd`.

I think it might be helpful for readers of the code if the invocations of `asyncStart` and `asyncEnd` that end up matching each other in the dynamic invocation sequence were also the ones that are located together syntactically. Anyway, that's good for reasoning about the sequence of events, but it isn't a requirement for passing the test run.

We can also have a structure that looks like `()()()()()`, as shown in 'async/async_test.dart'. The actual sequence of events in that case is probably `((((()))))` because the top-level statement list in `main` is executed without suspension, and then all the callbacks (function literals passed to `.then`) are executed after `main` has returned. Again, this works, but the actual sequencing of events is not necessarily self-explanatory.

Reply:
> > asyncEnd should never be called when there are more asyncStarts to be run in the program.

> I do think 'async/error_timing_test.dart' will do this, and this test succeeds on 79 out of 83 configurations at this time. I think the actual requirement is that the parenthesis structure is maintained. 

You're right, the rule I mentioned more strict than necessary. The parenthesis rule mostly makes sense to me, but

> We can also have a structure that looks like ()()()()(),

Here the first ')' will just end the test. There needs to be an outer parens like `(()()()()())`.

> @osa1 wrote that @mkustermann said:

To clarify, I wasn't quoting Martin. I just wrote my understanding from our discussion. All mistakes are mine :sweat_smile: .

---

I think we just can't make the `asyncStart`/`asyncEnd` have the same syntactic and runtime order while still testing everything we want to test. The example in my previous comment:

```dart
main()  {
    asyncStart();
    Future<int>.delayed(Duration(seconds: 1), () {
      ...
      asyncEnd();
      return 123;
    });
    Expect.equals(...);
}
```

I agree that it would be good if `asyncStart` and `end`s occured symmetrically and in the same lexical scope, but I'm not sure if that's possible in general.

Reply:
@osa1 wrote, in response to a remark from me:
>> We can also have a structure that looks like ()()()()(),
>
> Here the first ')' will just end the test. There needs to be an outer parens like (()()()()()).

Right, that's the reason why I said "looks like" and then continued to note that the actual sequencing in that test ('async/async_test.dart') is `((((()))))`, in spite of the fact that it looks like `()()()()()` syntactically. The point I was trying to make was that the syntax can look nice and still be misleading. In this case, 'async_test.dart' can succeed even though it doesn't syntactically have an extra pair of `asyncStart()`/`asyncEnd()` around the whole thing, but it just takes one `await` to break that.

Actually, we can have a `main` like this and get in trouble:

```dart
void main() {
  asyncStart();
  ... // Perform the actions 
  asyncEnd();
}
```

This test will fail with an 'asyncStart() was called even though we are done with testing' error if all the actions are delayed, because we will then invoke `asyncStart()` (the first one) at the beginning of main, then `asyncEnd()` (the one that is considered to mark the end of testing) at the end of `main`, and then a bunch of other `asyncStart()`/`asyncEnd()` pairs during the execution of various callbacks after the control has returned from `main`&mdash;except that the test will fail soon after `main` returns. This can easily happen, say because the actions are `async` functions that are called in `main` and whose returned futures are discarded.

I tend to think that the safe approach would be to aim for the following style, in all functions used by a test, with `main` as just one example:

```dart
// One approach seen in language tests.
main() {
  asyncStart();
  test().then((_) { asyncEnd(); });
}

// Alternatively.
Future<T> test() async {
  asyncStart();
  await anotherTest();
  asyncEnd();
}
```

So the idea is that `asyncStart()` occurs "synchronously" at the top level, and the corresponding `asyncEnd()` occurs in a callback to `.then`; or, equivalently, `asyncEnd()` occurs after code that `await`s all futures.

Reply:
I mean comparing of
```dart
void main() async {
  asyncStart();
  await test(); 
  asyncEnd();
}
```
vs
```dart
void main() async {
  await test(); 
}
```
I still think that `await` should discover all of the issues that may happen with `test()` and async quards won't give much advantages. If `test()` won't be executed, then shouldn't we expect a timeout in both cases (with and withoun guards)?